### PR TITLE
(Documentation-only) datatype.h

### PR DESCRIPTION
**Pull request by ajreynol**
This pull request documents some of the new features of classes in expr/datatype.h and cleans them to be in the new format.

It references a new refactoring issue #1344, which will affect how defined functions are represented in sygus datatypes.

Also note it references a paper draft "Datatypes with Shared Selectors", this documentation will be updated once this is published.

### CODE REVIEW

File: src/expr/datatype.h [hunk 1]
```diff
@@ -763,24 +847,191 @@ class CVC4_PUBLIC Datatype {
...
+   * bound variable list containing x and y.
+   */
   Expr getSygusVarList() const;
-  /** does it allow constants */
+  /** get sygus allow constants
+   *
+   * Does this sygus datatype allow constants?
+   * Notice that this is not a property of the
+   * constructors of this datatype. Instead, it is
+   * pan auxiliary flag (provided in the call
```
Comment (HanielB) [inline src/expr/datatype.h]: pan -> an

File: src/expr/datatype.h [hunk 2]
```diff
@@ -763,24 +847,191 @@ class CVC4_PUBLIC Datatype {
...
+   */
   Type getSygusType() const;
-  /** get sygus var list */
+
+  /** get sygus var list
+   * This gets the variable list of the function
+   * to synthesize using this sygus datatype.
+   * For example, if we are synthesizing a binary
+   * function f where solutions are of the form:
+   *   f = (\ xy. t[x,y])
```
Comment (aniemetz) [inline src/expr/datatype.h]: Is the '\\' here a typo? Missing quantifier?

File: src/expr/datatype.h [hunk 3]
```diff
@@ -763,24 +847,191 @@ class CVC4_PUBLIC Datatype {
...
-  /** get the evaluation function for this datatype for the deep embedding */
+  /** get sygus evaluation function
+   *
+   * This gets the evaluation function for this datatype
+   * for the deep embedding. This is a function of type:
+   *   D x T1 x ... x Tn -> T
+   * where:
+   *   D is the datatype type for this datatype,
+   *   T1...Tn are the types of the variables in
+   *     getSygusVarList(),
```
Comment (aniemetz) [inline src/expr/datatype.h]: Move to line above.

File: src/expr/datatype.h (hunk 2 above)
Comment (ajreynol) [reply src/expr/datatype.h]: This is a lambda, however, I realize I'm not using the notation consistently, I'll switch to "lambda xy"

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

**Pull request merged**